File: miscellaneous/union-find/3-weighted-union-find/main.go
package main

import "fmt"

/*
	Write		O(M)
	Find		O(logn)
	Union		O(N)
	Overall	O(MlogN)
*/

type UnionFind struct {
	Count int
	Ids   []int
	Caps  []int
}

func Constructor(n int) UnionFind {
	ids := []int{}
	caps := []int{}
	for i := 0; i < n; i++ {
		ids = append(ids, i)
		caps = append(caps, 1)
	}
	return UnionFind{n, ids, caps}
}

func (this *UnionFind) GetCount() int {
	return this.Count
}

// O(logN)
func (this *UnionFind) Find(key int) int {
	if key < len(this.Ids) {
		// loop to find to ultimate root
		cur := key
		for cur != this.Ids[cur] {
			cur = this.Ids[cur]
		}
		return cur
	}
	return -1
}

// union to the bigger tree, O(N)
func (this *UnionFind) Union(p int, q int) {
	if p < 0 || p+1 > len(this.Ids) || q < 0 || q+1 > len(this.Ids) {
		return
	}
	pId := this.Find(p)
	qId := this.Find(q)
	if pId == qId {
		return
	}

	if this.Caps[pId] < this.Caps[qId] {
		this.Ids[pId] = qId
		this.Caps[qId] += this.Caps[pId]
	} else {
		this.Ids[qId] = pId
		this.Caps[pId] += this.Caps[qId]
	}
	this.Count--
}

func (this *UnionFind) IsConnect(p int, q int) bool {
	if p < 0 || p+1 > len(this.Ids) || q < 0 || q+1 > len(this.Ids) {
		return false
	}
	return this.Find(p) == this.Find(q)
}

func main() {
	c := Constructor(10)
	c.Union(2, 3)
	c.Union(3, 4)
	fmt.Println(c.Ids)
	fmt.Println(c.Find(3))
	fmt.Println(c.Find(4))
	fmt.Println(c.IsConnect(2, 3))
	fmt.Println(c.IsConnect(3, 4))
	fmt.Println(c.GetCount())
}

File: miscellaneous/union-find/3-weighted-union-find/main.py
"""
    ref:
    - https://algs4.cs.princeton.edu/15uf/
"""


class UnionFind(object):
    def __init__(self, vertices):
        self.count = len(vertices)
        self.roots = {}
        self.caps = {}
        for vertex in vertices:
            self.roots[vertex] = vertex
            self.caps[vertex] = 1

    def getCount(self):
        return self.count

    def find(self, key):
        # loop to find to the ultimate root
        if key not in self.roots:
            return None
        cur = key
        while cur != self.roots[cur]:
            cur = self.roots[cur]
        return cur

    def union(self, p, q):
        if p not in self.roots or q not in self.roots:
            return
        pId = self.find(p)
        qId = self.find(q)
        if pId == qId:
            return
        # attach to the larger tree
        if self.caps[pId] < self.caps[qId]:
            self.roots[pId] = qId
            self.caps[qId] += self.caps[pId]
        else:
            self.roots[qId] = pId
            self.caps[pId] += self.caps[qId]
        self.count -= 1

    def isConnect(self, p, q):
        if p not in self.roots or q not in self.roots:
            return None
        return self.find(p) == self.find(q)


nodes = ['a', 'b', 'c', 'd', 'e']
uf = UnionFind(nodes)

uf.union('a', 'b')
uf.union('c', 'd')
uf.union('d', 'e')

ht = {}
for node in nodes:
    root = uf.find(node)
    if root in ht:
        ht[root].append(node)
    else:
        ht[root] = [node]

print(ht)

largestGroup = []
for key in ht:
    if len(ht[key]) > len(largestGroup):
        largestGroup = ht[key]
print(largestGroup)
